fido2-core/src/main/java/com/linecorp/line/auth/fido/fido2/server/service/ResponseServiceImpl.java

```diff
@@ -151,26 +152,14 @@ protected RegisterCredentialResult createRegisterCredentialResult(AuthenticatorD
 
     @Override
     protected void checkOrigin(URI originFromClientData, URI originFromRp) {
-        final String ANDROID_FACET_SCHEME = "android";
-        final String IOS_FACET_SCHEME = "ios";
+        List<String> configuredOrigins = appOriginService.getOrigins(originFromRp.toString());
 
-        if (originFromClientData.getScheme().equals(ANDROID_FACET_SCHEME) ||
-                originFromClientData.getScheme().equals(IOS_FACET_SCHEME)) {
-            //app case
-            List<String> appOriginList = appOriginService.getOrigins(originFromRp.toString());
-
-            if (!appOriginList.contains(originFromClientData.toString())) {
-                throw new FIDO2ServerRuntimeException(InternalErrorCode.ORIGIN_NOT_MATCHED,
-                        "Client facet origin: " + originFromClientData + ", App Origin List: " + appOriginList);
-            }
-
-        } else {
-            // web case
-            if (!originFromRp.toString().equals(originFromClientData.toString())) {
-                throw new FIDO2ServerRuntimeException(InternalErrorCode.ORIGIN_NOT_MATCHED,
-                        "From collected data: " + originFromClientData + ", From request param: " + originFromRp);
-            }
+        if (OriginValidationHelper.isAppFacet(originFromClientData)) {
+            OriginValidationHelper.validateAppFacet(originFromClientData, configuredOrigins);
+            return;
         }
+
+        OriginValidationHelper.validateWeb(originFromClientData, originFromRp, configuredOrigins);
```

Comment:
Maybe we can create `OriginValidationService` and provide this logic as method `validate`.
In `OriginValidationService`, we can move `OriginValidationHelper` logics as private method. How do you think about this direction?

Reply:
@gibeom-gwon 

Thanks for the suggestion! How about we merge `AppOriginService`  and `OriginValidationHelper` into one OriginValidationService that handles both origin lookups and validation?